py/asmrv32: Add support for RV32 Zba opcodes.

agatti opened the pull request:
### Summary

This PR adds support for RV32 [Zba](https://github.com/riscv/riscv-isa-manual/blob/main/src/b-st-ext.adoc#zba) address generation opcodes to several places in MicroPython (scaled addition operation used for calculating halfword and word offsets):

* the native emitter will emit Zba opcodes if configured to do so for Viper/Native 16- and 32-bits register-indexed load/store operations
* the inline assembler is now able to emit `SH1ADD`, `SH2ADD`, and `SH3ADD` opcodes
* the `rp2` port will emit `Zba` opcodes at runtime for RP2350 firmwares when built using the RISCV variant
* `mpy-cross` now has an extra implementation-dependent option to enable generation of `Zba` opcodes when pre-compiling native or viper code into a MPY file

~~With these changes, each load/store operation that can use these opcodes will be 2 bytes shorter.~~

With these changes, all non-byte load/store operations have been shortened by two bytes, and operations that can use these opcodes will be one instruction shorter (fewer CPU cycles and fewer register stalls).

### Testing

The inline assembler changes were tested on QEMU (to check they work), and on an ESP32C3 (to check whether the appropriate tests are skipped if those opcodes are not supported).

Runtime changes were tested on QEMU, built with and without `MICROPY_EMIT_RV32_ZBA` being set, running the full RV32 test suite.

`mpy-cross` changes were tested by looking at the emitted output code of viper ptr tests through a disassembler; both with and without `-X riscv=zba` in `mpy-cross` command line.

### Trade-offs and Alternatives

There's a fixed 4 bytes (or 8 bytes on 64 bit targets) RAM overhead to let the dynamic compiler settings structure hold a pointer to an emitter-specific options structure.

This mechanism, in theory, could lead to the deprecation of the various Armv7 sub-types, as they could be command line switches instead.

agatti committed:
py/mpstate: Make it possible for mpy-cross to set emitter options.

This commit introduces a way for mpy-cross to pass a set of options to
the chosen emitter.

This is achieved by adding an opaque pointer to the dynamic compiler
state structure that is only accessed by emitters that have a need to
receive options from mpy-cross when generating code.  That's a way to
make this feature possible without breaking any existing code or
emitter, and without re-engineering the compiler entry point function
(together with passing the options struct downstream until it's time to
emit code).

The main use case for this is letting the RV32 emitter know which
optional extensions it can generate code with, to be able to emit better
suited code for the platform in use.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
py/emitinlinerv32: Add Zba opcodes to the inline assembler.

This commit adds support for Zba opcodes to the RV32 inline assembler.

Three new opcodes were added, SH1ADD, SH2ADD, and SH3ADD, which performs
a scaled addition (by 1, 2, or 3 bits respectively).  At the moment only
qemu's VIRT_RV32 and rp2's RPI_PICO2/RPI_PICO2_W ports support these
opcodes (the latter only when using the RISCV variant).

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
rp2/mpconfigport: Enable Zba opcodes in RISC-V mode.

This commit enables generation of Zba opcodes by the native emitter
for the Pico2, as its RISC-V implementation supports both of those
extensions (see section 3.8 of the RP2350 datasheet).

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
mpy-cross/main: Add support for RV32 Zba opcodes.

This commit adds a new command line switch to inform the RV32 emitter to
use Zba opcodes in its output.

A new implementation-specific option was introduced, called
"-march-flags", that will contain a list of additional
architecture-specific flags to pass to the chosen native emitter
implementation.

At the moment only the RV32 emitter can make use of this command line
facility: if the architecture flags string equals to "zba"
(case-sensitive), then the native emitter will emit Zba opcodes if it
has a chance to do so.

At the moment there is no check on whether additional architecture flags
using to build a MPY file are compatible with the target the output code
is run on, so use this with caution.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti commented:
Weird, the `sanitize_address` test failure doesn't seem related to my changes - my code doesn't touch threading at all (the failing test is `thread/stress_recurse.py`).

agatti commented:
That could work, although extra care should be taken when displaying the available extensions from `mpy-cross`'s help to make sure those extensions are discoverable.

As a thought experiment, let's say that for MPY v7 you want to coalesce all ArmV7 targets into one single platform identifier and use this backend option handling mechanism for the thumb emitter (that may simplify the loader and natmod tooling, and maybe even save some space on the firmware side [1]).  If you follow the same rationale then you'd have to add "-mfloat-abi" and "-mfpu" command line arguments to `mpy-cross` to follow GCC.

I'd personally go for something more platform-neutral, like "-mfeatures" for example.  This leaves the door open for other platforms' configuration options in the future.

[1] And after you coalesce the armv7 platforms, if you run an armv7emdp natmod on a Netduino then you'd have the same protections as if you're running a Zba-enabled natmod on an ESP32C3 with this PR merged in :)

dpgeorge commented on ports/rp2/CMakeLists.txt:
```diff
@@ -487,7 +487,16 @@ list(APPEND MICROPY_SOURCE_QSTR
+if (${PICO_PLATFORM} STREQUAL "rp2040")
+    set(MICROPY_CROSS_FLAGS "-march=armv6m")
+elseif (${PICO_PLATFORM} STREQUAL "rp2350-arm-s")
+    set(MICROPY_CROSS_FLAGS "-march=armv7emsp")
```
RP2350 uses softfp calling conventions (unfortunately).  But that's something which is out of scope for this PR.

dpgeorge commented:
> If that's OK with you I can work on this

Yes, that's fine with me.  Although I have some further suggestions:
1. We can get this PR merged much earlier by following my option (3) above and only enable Zba support in the on-device compiler/emitter, and disable it in mpy-cross (therefore no need to encode anything in the .mpy file).  That gets most of the functionality in (all of the relevant native emitter / assembler code) and doesn't need any further discussion/design.  (Or even keep the new mpy-cross `-march-flags` option and users should be aware that it may generate code which is incompatible with their target, because there are no checks for the flags at this stage.)
2. Similarly, we can merge the RV64 qemu PR #15931 without float support.
3. I have some extra things I want to add to native .mpy files (eg to properly support bss and data sections) and that needs some additional bytes of information in the .mpy file.  So I would suggest we use the mpy `header[2]` byte as follows: bit pattern `0b00aaaa11` is the existing format, but the loader also supports bit pattern `0b01aaaa11 which means that there are some extra bytes that follow in the .mpy file (or even that the rest of the .mpy file is totally different).  Then we are free to add whatever extra fields we need, eg I would suggest a variable length encoded uint for the arch flags, to allow arbitrary room for growth in the number of flags.

agatti committed:
qemu: Enable Zba opcodes for the VIRT_RV32 board.

This commit enables the usage of the Zba address generation opcodes in
the QEMU port when targeting the "VIRT_RV32" board, which emulates a
generic RV32 target.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

dpgeorge commented on mpy-cross/main.c:
```diff
@@ -131,6 +139,8 @@ static int usage(char **argv) {
         "-msmall-int-bits=number : set the maximum bits used to encode a small-int\n"
         "-march=<arch> : set architecture for native emitter;\n"
         "                x86, x64, armv6, armv6m, armv7m, armv7em, armv7emsp, armv7emdp, xtensa, xtensawin, rv32imc, host, debug\n"
+        "-march-flags=<flags> : set architecture-specific flags (OUTPUT FILE MAY NOT WORK ON ALL TARGET BOARDS)\n"
```
Suggest using just "TARGETS" instead of "TARGET BOARDS".

dpgeorge commented on mpy-cross/main.c:
```diff
@@ -350,6 +364,29 @@ MP_NOINLINE int main_(int argc, char **argv) {
+        }
+        mp_printf(&mp_stderr_print,
+            "Warning: Using architecture-specific flags may create a MPY file whose code\n"
+            "         won't run on all target boards!\n"
```
"targets" instead of "target boards"

agatti replied:
This should have been addressed in the latest version of the PR.

dpgeorge commented on py/emitinlinerv32.c:
```diff
@@ -37,6 +37,23 @@
+#endif
+
+#define ALLOWED_EXTENSIONS (EXTENSION_ZBA_FRAGMENT)
+static uint32_t emit_inline_rv32_allowed_extensions = 0;
```
It looks like this variable is only needed for `MICROPY_DYNAMIC_COMPILER`, and so I suggest guarding it in such a `#if MICROPY_DYNAMIC_COMPILER`.

Actually, that's still a bit messy and essentially duplicates the `asm_rv32_allow_zba_opcodes()` function added to `asmrv32.c`.

So my suggestion would be to follow and extend how Thumb does it with the public `asm_thumb_allow_armv7m()` function (which is used by `asmthumb.c` and `emitinlinethumb.c`, and even `emitnative.c`):
- standardise in `asmrv32h` a public set of bit fields (eg enum) for supported RV32 extensions
- change the `asm_rv32_backend_options_t` struct to have a `uint8_t rv32_ext;` entry which holds a bit mask of enabled extensions (instead of bools for each flag)
- expose a public function in `asmrv32.h` (maybe `static inline`) called something like `asm_rv32_allowed_extensions()` which is similar to how `asm_rv32_allow_zba_opcodes` is currently implemented and can replace (or be used to implement) that function

Eg in `asmrv32.h`:
```c
enum {
   RV32_EXT_ZBA = 1 << 0,
   ...
};

static inline uint8_t asm_rv32_allowed_extensions(void) {
    uint8_t ext = 0;
    #if MICROPY_DYNAMIC_COMPILER
    if (mp_dynamic_compiler.backend_options != NULL) {
        ext |= ((asm_rv32_backend_options_t *)mp_dynamic_compiler.backend_options)->ext;
    }
    #else
    #if MICROPY_EMIT_RV32_ZBA
    ext |= RV32_EXT_ZBA;
    #endif
    #endif
    return ext;
}

static inline bool asm_rv32_allow_zba_opcodes(void) {
    return asm_rv32_allowed_extensions() & RV32_EXT_ZBA;
}

dpgeorge commented on py/emitinlinerv32.c:
```diff
@@ -220,6 +245,9 @@ typedef void (*call_i_t)(asm_rv32_t *state, mp_int_t immediate);
 typedef void (*call_n_t)(asm_rv32_t *state);
 
+#define EXT_NIL (0)
+#define EXT_ZBA (1 << (EXTENSION_BIT_ZBA))
```
As suggested above, these could be moved to `asmrv32.h` and reused by mpy-cross etc.

agatti replied:
This should be addressed in the latest version of the PR.

dpgeorge commented on py/emitinlinerv32.c:
```diff
@@ -785,10 +817,11 @@ static void emit_inline_rv32_opcode(emit_inline_asm_t *emit, qstr opcode, mp_uin
-                MP_ERROR_TEXT("unknown RV32 instruction '%q'"), opcode));
+    if (!opcode_data ||
+        (((opcode_data->required_extensions & ALLOWED_EXTENSIONS) != opcode_data->required_extensions) &&
+         ((opcode_data->required_extensions & emit_inline_rv32_allowed_extensions) != opcode_data->required_extensions))) {
```
As suggested above, this could be simplified to `opcode_data->required_extensions & asm_rv32_allowed_extensions() != opcode_data->required_extensions`.

agatti committed:
py/asmrv32: Use RV32 Zba opcodes if possible.

This commit adds optional support for selected Zba opcodes (address
generation) to speed up Viper and native code generation on MCUs where
those opcodes are supported (namely RP2350).

Right now support for these opcodes is opt-in, as extension detection
granularity on the RISC-V platform is still a bit in flux.  Relying on
the 'B' bit in the MISA register may yield both false positives and
false negatives depending on the RISC-V implementation the check runs
on.

As a side-effect of Zba support, regular non-byte load/stores have been
made shorter by two bytes.  Whilst this makes code using Zba take up the
same space as non-Zba code, the former will still be faster as it will
have to process just one instruction instead of two, without stalling
registers between the shift and the addition needed to compute the final
offset.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

dpgeorge reviewed (approved):
Thanks for following my suggestion and updating the code.  It looks good now.

And thanks overall for persisting with this PR, it now sets the direction to add arch-specific extensions to the emitter.  And having zba for RV32 targets is a nice enhancement.
